### PR TITLE
restore performance improvements (part2)

### PR DESCRIPTION
**Pull request by snopoke**
dependent on https://github.com/dimagi/commcare-hq/pull/5072

2 main changes in this PR
## TL/DR

For  a user with 500 cases:

_Only wrap if necessary_
1st restore = 172s (110s wrapping)
2nd restore (sync) using data from synclog = 23s (15s filter_cases_modified_elsewhere_since_sync)
2nd restore (sync) fetching data from couch = 23s (14s filter_cases_modified_elsewhere_since_sync)

_Wrap everything_
1st = 173s (112s fetch and wrap)
2nd = 133s (112s fetch and wrap)
## Don't wrap docs if you don't have to

The idea here is to do all the processing on raw cases and only wrap the ones that need to be sent down to the user right at the end.

Profiling this for a user with 500 cases, the full restore took 172 seconds to generate of which 110 seconds was spent wrapping the docs. A further improvement could be gained here if we didn't need to wrap the docs at all which requires changes to this: https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/casexml/apps/phone/xml.py#L33
## Don't fetch docs from couch if you don't need to

This only works if the user is not part of any case sharing groups. The idea here is that all the calculations in the restore only require the case_id and case indices. Most of these (if not all) are already available in the SyncLog so just use those instead of fetching them from couch. At the end of the process when we know what cases we need to send to the user we fetch them.

This didn't produce as much improvement as I expected but I put that down to running locally where fetching from couch is fast.

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 8 errors:

* corehq/ex-submodules/casexml/apps/phone/caselogic.py, line 400 - E501 line too long (116 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/caselogic.py, line 402 - E501 line too long (138 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/caselogic.py, line 421 - E501 line too long (118 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/tests/__init__.py, line 7 - F401 'BatchRestoreTests' imported but unused
* corehq/ex-submodules/casexml/apps/phone/tests/__init__.py, line 7 - F401 'MultiUserSyncTestBatched' imported but unused
* corehq/ex-submodules/casexml/apps/phone/tests/__init__.py, line 7 - F401 'OtaRestoreTestBatched' imported but unused
* corehq/ex-submodules/casexml/apps/phone/tests/__init__.py, line 7 - F401 'SyncTokenUpdateTestBatched' imported but unused
* corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py, line 44 - W291 trailing whitespace

**Comment by snopoke** [comment]
@czue I re-looked at this and the scenario you described won't be a problem since I always load all the case IDs from couch so any new ones still get included.

Added a unit test to verify.

File: corehq/ex-submodules/casexml/apps/phone/tests/__init__.py [hunk 1]
```diff
@@ -4,6 +4,12 @@
     from .test_state_hash import *
     from .test_sync_logs import *
     from .test_sync_mode import *
+    from .test_batched_mode import (
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/__init__.py]: F401 'BatchRestoreTests' imported but unused

Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/__init__.py]: F401 'MultiUserSyncTestBatched' imported but unused

Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/__init__.py]: F401 'OtaRestoreTestBatched' imported but unused

Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/__init__.py]: F401 'SyncTokenUpdateTestBatched' imported but unused

Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/__init__.py]: F401 'BatchRestoreTests' imported but unused

Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/__init__.py]: F401 'MultiUserSyncTestBatched' imported but unused

Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/__init__.py]: F401 'OtaRestoreTestBatched' imported but unused

Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/__init__.py]: F401 'SyncTokenUpdateTestBatched' imported but unused

**Comment by snopoke** [comment]
@czue :white_check_mark:

**Comment by czue** [comment]
thanks @biyeun and @snopoke. sorry was slow to review!
